source/common/grpc/typed_async_client.h

```diff
@@ -0,0 +1,145 @@
+  bool operator==(AsyncStream* stream) const { return stream_ == stream; }
+  bool operator!=(AsyncStream* stream) const { return stream_ != stream; }
+
+private:
+  template <typename Request> friend class TypedAsyncStream;
+  AsyncStream* stream_{};
+  bool is_grpc_header_required_{};
+};
+
+template <typename Request> class TypedAsyncStream : public UntypedAsyncStream {
+public:
+  TypedAsyncStream() {}
+  TypedAsyncStream(AsyncStream* stream, bool is_grpc_header_required)
+      : UntypedAsyncStream(stream, is_grpc_header_required) {}
+  TypedAsyncStream(UntypedAsyncStream&& other)
+      : TypedAsyncStream(std::move(other.stream_), other.is_grpc_header_required_) {}
+  void sendMessageTyped(const Request& request, bool end_stream) {
+    sendMessageUntyped(request, end_stream);
+  }
+  TypedAsyncStream* operator->() { return this; }
+  TypedAsyncStream<Request> operator=(AsyncStream* stream) {
+    stream_ = stream;
+    return *this;
+  }
+};
+
+/**
+ * Convenience subclasses for AsyncRequestCallbacks.
+ */
+class UntypedAsyncRequestCallbacks : public AsyncRequestCallbacks {
+public:
+  virtual ~UntypedAsyncRequestCallbacks() {}
+  virtual void onSuccessUntyped(ProtobufTypes::MessagePtr&& response, Tracing::Span& span) PURE;
+
+protected:
+  void onSuccess(Buffer::InstancePtr&& response, Tracing::Span& span) override;
+  virtual ProtobufTypes::MessagePtr createEmptyResponse() { NOT_REACHED_GCOVR_EXCL_LINE; }
+};
+
+template <typename Response>
+class TypedAsyncRequestCallbacks : public UntypedAsyncRequestCallbacks {
+public:
+  virtual ~TypedAsyncRequestCallbacks() {}
+  virtual void onSuccessTyped(std::unique_ptr<Response>&& response, Tracing::Span& span) PURE;
+
+private:
+  ProtobufTypes::MessagePtr createEmptyResponse() override { return std::make_unique<Response>(); }
+  void onSuccessUntyped(ProtobufTypes::MessagePtr&& response, Tracing::Span& span) override {
+    onSuccessTyped(std::unique_ptr<Response>(dynamic_cast<Response*>(response.release())), span);
+  }
+};
+
+/**
+ * Convenience subclasses for AsyncStreamCallbacks.
+ */
+class UntypedAsyncStreamCallbacks : public AsyncStreamCallbacks {
+public:
+  virtual ~UntypedAsyncStreamCallbacks() {}
+  virtual void onReceiveMessageUntyped(ProtobufTypes::MessagePtr&& response) PURE;
+
+protected:
+  virtual ProtobufTypes::MessagePtr createEmptyResponse() { NOT_REACHED_GCOVR_EXCL_LINE; }
+  bool onReceiveMessage(Buffer::InstancePtr&& response) override;
+};
+
+template <typename Response> class TypedAsyncStreamCallbacks : public UntypedAsyncStreamCallbacks {
+public:
+  virtual ~TypedAsyncStreamCallbacks() {}
+  virtual void onReceiveMessageTyped(std::unique_ptr<Response>&& message) PURE;
+
+private:
+  ProtobufTypes::MessagePtr createEmptyResponse() override { return std::make_unique<Response>(); }
+  void onReceiveMessageUntyped(ProtobufTypes::MessagePtr&& response) override {
+    onReceiveMessageTyped(std::unique_ptr<Response>(dynamic_cast<Response*>(response.release())));
+  }
+};
```

Comment:
Same for all these dual untyped/typed wrappers.

Reply:
Done.  PTAL